File: flexinput/src/main/java/com/lytefast/flexinput/fragment/CameraFragment.java
package com.lytefast.flexinput.fragment;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.provider.MediaStore;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.Toast;

import com.google.android.cameraview.CameraView;
import com.lytefast.flexinput.FlexInputCoordinator;
import com.lytefast.flexinput.R;
import com.lytefast.flexinput.R2;
import com.lytefast.flexinput.utils.FileUtils;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;



/**
 * {@link Fragment} that allows the user to take a live photo directly from the camera.
 *
 * @author Sam Shih
 */
public class CameraFragment extends PermissionsFragment {
  private static final String[] REQUIRED_PERMISSIONS = {
      Manifest.permission.WRITE_EXTERNAL_STORAGE,
      Manifest.permission.READ_EXTERNAL_STORAGE,
      Manifest.permission.CAMERA};

  @CameraView.Flash
  private static final int[] FLASH_STATE_CYCLE_LIST = {
      CameraView.FLASH_AUTO,
      CameraView.FLASH_ON,
      CameraView.FLASH_OFF};

  private static final String TAG = CameraFragment.class.getCanonicalName();

  public static final int REQUEST_IMAGE_CAPTURE = 4567;

  @BindView(R2.id.camera_view) CameraView cameraView;
  @BindView(R2.id.camera_container) View cameraContainer;
  @BindView(R2.id.permissions_container) FrameLayout permissionsContainer;
  private Unbinder unbinder;

  private FlexInputCoordinator flexInputCoordinator;

  /**
   * Temporary holder for when we intent to the camera. This is used because the resulting intent
   * doesn't return any data if you set the output file in the request.
   */
  private File photoFile;

  @Override
  public void onCreate(@Nullable final Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setRetainInstance(true);
  }

  @Nullable
  @Override
  public View onCreateView(final LayoutInflater inflater, @Nullable final ViewGroup container, @Nullable final Bundle savedInstanceState) {
    final Fragment targetFragment = getParentFragment().getTargetFragment();
    if (targetFragment instanceof FlexInputCoordinator) {
      this.flexInputCoordinator = (FlexInputCoordinator) targetFragment;
    }

    View rootView = inflater.inflate(R.layout.fragment_camera, container, false);
    unbinder = ButterKnife.bind(this, rootView);

    cameraView.addCallback(cameraCallback);

    tryStartCamera();
    return rootView;
  }

  @Override
  public void onResume() {
    super.onResume();

    if (!getContext().getPackageManager().hasSystemFeature(PackageManager.FEATURE_CAMERA)
        || !hasPermissions(REQUIRED_PERMISSIONS)) {
      cameraContainer.setVisibility(View.GONE);
      permissionsContainer.setVisibility(View.VISIBLE);

      initPermissionsView(permissionsContainer);

      return;  // No camera detected. just chill
    }
    cameraContainer.setVisibility(View.VISIBLE);
    permissionsContainer.setVisibility(View.GONE);

    // Delayed restart since we are coming back from camera, and for some reason the API
    // isn't fast enough to acknowledge the other activity closed the camera.
    cameraView.postDelayed(new Runnable() {
      @Override
      public void run() {
        if (cameraView != null) {
          tryStartCamera();
        }
      }
    }, 350);
  }

  protected void initPermissionsView(final FrameLayout permissionsContainer) {
    View view = LayoutInflater.from(permissionsContainer.getContext())
        .inflate(R.layout.view_camera_permissions, permissionsContainer, true);
    View reqBtn = ButterKnife.findById(view, R.id.permissions_req_btn);
    reqBtn.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(final View v) {
        requestPermissionClick();
      }
    });
  }

  /**
   * Some cameras don't properly set the {@link android.hardware.Camera.CameraInfo#facing} value.
   * So here, if we fail, just try getting the first front facing camera.
   */
  private void tryStartCamera() {
    try {
      if (cameraView.isCameraOpened()) {
        cameraView.stop();
      }
      cameraView.start();
    } catch (Exception e) {
      Log.w(TAG, "Camera could not be loaded, try front facing camera", e);

      try {
        cameraView.setFacing(CameraView.FACING_FRONT);
        cameraView.start();
      } catch (Exception ex) {
        Log.e(TAG, "Camera could not be loaded", e);
      }
    }

    ImageView flashBtn = ButterKnife.findById(cameraContainer, R.id.camera_flash_btn);
    setFlash(flashBtn, CameraView.FLASH_AUTO);
  }

  @Override
  public void onPause() {
    cameraView.stop();
    super.onPause();
  }

  @Override
  public void onDestroyView() {
    unbinder.unbind();
    super.onDestroyView();
  }

  protected void requestPermissionClick() {
    requestPermissions(new PermissionsResultCallback() {
      @Override
      public void granted() {
        cameraView.post(new Runnable() {
          @Override
          public void run() {
            // #onResume will take care of this for us
          }
        });
      }

      @Override
      public void denied() {
      }
    }, REQUIRED_PERMISSIONS);
  }

  @OnClick(R2.id.camera_flash_btn)
  void onCameraFlashClick(ImageView flashBtn) {
    final int currentFlashState = cameraView.getFlash();
    int currentStateIndex = -1;
    for (int i = 0; i < FLASH_STATE_CYCLE_LIST.length; i++) {
      if (currentFlashState == FLASH_STATE_CYCLE_LIST[i]) {
        currentStateIndex = i;
        break;
      }
    }

    final int newStateIndex = (currentStateIndex + 1) % FLASH_STATE_CYCLE_LIST.length;
    setFlash(flashBtn, FLASH_STATE_CYCLE_LIST[newStateIndex]);
  }

  @OnClick(R2.id.take_photo_btn)
  void onTakePhotoClick() {
    if (cameraView.isCameraOpened()) {
      cameraView.takePicture();
    }
  }

  @OnClick(R2.id.launch_camera_btn)
  void onLaunchCameraClick() {
    cameraView.stop();

    photoFile = flexInputCoordinator.getFileManager().newImageFile();
    Uri photoUri =  flexInputCoordinator.getFileManager().toFileProviderUri(getContext(), photoFile);
    Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        .putExtra(MediaStore.EXTRA_OUTPUT, photoUri)
        .addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION);

    if (takePictureIntent.resolveActivity(getContext().getPackageManager()) != null) {
      grantWriteAccessToURI(getContext(), takePictureIntent, photoUri);
      startActivityForResult(takePictureIntent, REQUEST_IMAGE_CAPTURE);
      // TODO need to handle the result: 1) save thumbnail, 2) call #addToMediaStore
    }
  }

  @Override
  public void onActivityResult(final int requestCode, final int resultCode, final Intent data) {
    super.onActivityResult(requestCode, resultCode, data);
    if (REQUEST_IMAGE_CAPTURE != requestCode) {
      return;
    }
    if (Activity.RESULT_CANCELED == resultCode) {
      // Do nothing
    } else if (Activity.RESULT_OK != resultCode) {
      Toast.makeText(getContext(), R.string.camera_intent_result_error, Toast.LENGTH_SHORT).show();
      if (photoFile != null) {
        photoFile.delete();  // cleanup
      }
    } else if (photoFile != null) {
      flexInputCoordinator.addExternalAttachment(FileUtils.toAttachment(photoFile));
      return;
    }
  }

  private final CameraView.Callback cameraCallback = new CameraView.Callback() {

        @Override
        public void onCameraOpened(CameraView cameraView) {
            Log.d(TAG, "onCameraOpened");
        }

        @Override
        public void onCameraClosed(CameraView cameraView) {
            Log.d(TAG, "onCameraClosed");
        }

        @Override
        public void onPictureTaken(CameraView cameraView, final byte[] data) {
            Log.d(TAG, "onPictureTaken " + data.length);
            Toast.makeText(cameraView.getContext(), "Picture saved", Toast.LENGTH_SHORT)
                    .show();

          AsyncTask.execute(new Runnable() {
            @Override
            public void run() {
              File file = flexInputCoordinator.getFileManager().newImageFile();
              OutputStream os = null;
              try {
                os = new FileOutputStream(file);
                os.write(data);
                os.close();

                addToMediaStore(file);
                flexInputCoordinator.addExternalAttachment(FileUtils.toAttachment(file));
              } catch (IOException e) {
                Log.w(TAG, "Cannot write to " + file, e);
              } finally {
                if (os != null) {
                  try {
                    os.close();
                  } catch (IOException e) {
                    // Ignore
                  }
                }
              }
            }
          });
        }
  };

  private void setFlash(final ImageView btn, @CameraView.Flash final int newFlashState) {
    if (cameraView.getFlash() == newFlashState) {
      return;
    }

    @DrawableRes final int flashImage;
    @StringRes final int flashMsg;
    switch(newFlashState) {
      case CameraView.FLASH_ON:
        flashMsg = R.string.flash_on;
        flashImage = R.drawable.ic_flash_on_24dp;
        break;
      case CameraView.FLASH_OFF:
        flashMsg = R.string.flash_off;
        flashImage = R.drawable.ic_flash_off_24dp;
        break;
      default:
        flashMsg = R.string.flash_auto;
        flashImage = R.drawable.ic_flash_auto_24dp;
        break;
    }

    cameraView.setFlash(newFlashState);
    Toast.makeText(getContext(), flashMsg, Toast.LENGTH_SHORT).show();
    btn.setImageResource(flashImage);
  }

  private void addToMediaStore(final File photo) {
    Intent mediaScanIntent = new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE, Uri.fromFile(photo));
    getContext().sendBroadcast(mediaScanIntent);
    Log.d(TAG, "Photo added to MediaStore: " + photo.getName());
  }

  /**
   * This is a hack to allow the file provider API to still
   * work on older API versions.
   *
   * http://bit.ly/2iC4bUJ
   */
  private static void grantWriteAccessToURI(final @NonNull Context context,
                                            final @NonNull Intent intent,
                                            final @NonNull Uri uri) {
    final List<ResolveInfo> resInfoList = context
        .getPackageManager()
        .queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY);

    for (ResolveInfo resolveInfo : resInfoList) {
      final String packageName = resolveInfo.activityInfo.packageName;
      final int mode = Intent.FLAG_GRANT_WRITE_URI_PERMISSION | Intent.FLAG_GRANT_READ_URI_PERMISSION;

      context.grantUriPermission(packageName, uri, mode);
    }
  }
}

File: flexinput/src/main/java/com/lytefast/flexinput/fragment/CameraFragment.kt
package com.lytefast.flexinput.fragment

import android.Manifest
import android.annotation.TargetApi
import android.app.Activity
import android.content.Context
import android.content.Context.CAMERA_SERVICE
import android.content.Intent
import android.content.pm.PackageManager
import android.hardware.Camera
import android.hardware.camera2.CameraManager
import android.net.Uri
import android.os.AsyncTask
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.preference.PreferenceManager
import android.provider.MediaStore
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.Toast
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.annotation.UiThread
import androidx.fragment.app.Fragment
import com.camerakit.CameraKit
import com.camerakit.CameraKitView
import com.lytefast.flexinput.FlexInputCoordinator
import com.lytefast.flexinput.R
import com.lytefast.flexinput.utils.FileUtils.toAttachment
import java.io.File
import java.io.FileOutputStream
import java.io.IOException


/**
 * [Fragment] that allows the user to take a live photo directly from the camera.
 *
 * @author Sam Shih
 */
@Suppress("MemberVisibilityCanBePrivate")
open class CameraFragment : PermissionsFragment() {

  protected lateinit var cameraContainer: View
  protected lateinit var cameraView: CameraKitView
  protected lateinit var permissionsContainer: FrameLayout
  protected lateinit var cameraFacingBtn: ImageView

  private var flexInputCoordinator: FlexInputCoordinator<Any>? = null

  private lateinit var handler: Handler

  /**
   * Temporary holder for when we intent to the camera. This is used because the resulting intent
   * doesn't return any data if you set the output file in the request.
   */
  private var photoFile: File? = null

  private var flashState: Int
    @UiThread
    get() = with(PreferenceManager.getDefaultSharedPreferences(requireContext())) {
      getInt(PREF_FLASH_STATE, CameraKit.FLASH_AUTO)
    }
    @UiThread
    set(value) = with(PreferenceManager.getDefaultSharedPreferences(requireContext())) {
      edit()
          .putInt(PREF_FLASH_STATE, value)
          .apply()
    }

  var onFlashChanged: (flashBtn: View, newState: Int) -> Unit = { flashBtn, newState ->
    @StringRes val messageStringRes = when (newState) {
      CameraKit.FLASH_ON -> R.string.flash_on
      CameraKit.FLASH_OFF -> R.string.flash_off
//      CameraKit.FLASH_AUTO,
      else -> R.string.flash_auto
    }

    Toast.makeText(flashBtn.context, messageStringRes, Toast.LENGTH_SHORT).show()
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    retainInstance = true

    handler = Handler()
  }

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
    @Suppress("UNCHECKED_CAST")
    this.flexInputCoordinator = parentFragment?.parentFragment as? FlexInputCoordinator<Any>

    return inflater.inflate(R.layout.fragment_camera, container, false)
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)

    view.apply {
      permissionsContainer = findViewById(R.id.permissions_container)
      cameraContainer = findViewById(R.id.camera_container)

      cameraView = findViewById(R.id.camera_view)
      cameraView.errorListener = cameraErrorListener
      cameraView.cameraListener = cameraListener

      findViewById<View>(R.id.take_photo_btn)
          ?.setOnClickListener { cameraView.onTakePhotoClick() }
      findViewById<View>(R.id.launch_camera_btn)
          ?.setOnClickListener { onLaunchCameraClick() }

      val flashBtn = findViewById<ImageView>(R.id.camera_flash_btn)
      flashBtn?.setOnClickListener { onCameraFlashClick(it as ImageView) }
      setFlash(flashBtn, flashState, notify = false)

      cameraFacingBtn = findViewById(R.id.camera_facing_btn)
      cameraFacingBtn.setOnClickListener { onCameraFacingClick() }

      if (isSingleCamera) {
        cameraFacingBtn.visibility = View.GONE
      }
    }
  }

  private val isSingleCamera by lazy {
    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    when (Build.VERSION.SDK_INT) {
      in 0..Build.VERSION_CODES.KITKAT_WATCH -> Camera.getNumberOfCameras() == 1
      else -> {
        (context!!.getSystemService(CAMERA_SERVICE) as CameraManager)
            .cameraIdList.size == 1
      }
    }
  }

  override fun onStart() {
    super.onStart()
    cameraView.onStart()
  }

  override fun onResume() {
    super.onResume()

    if (!requireContext().packageManager.hasSystemFeature(PackageManager.FEATURE_CAMERA)
        || !hasPermissions(*REQUIRED_PERMISSIONS)) {
      cameraContainer.visibility = View.GONE
      permissionsContainer.also {
        it.visibility = View.VISIBLE
        if (it.childCount == 0) {
          initPermissionsView(it)
        }
      }

      return   // No camera detected. just chill
    }
    cameraContainer.visibility = View.VISIBLE
    permissionsContainer.visibility = View.GONE

    cameraView.onResume()
  }

  protected open fun initPermissionsView(permissionsContainer: FrameLayout) {
    val view = LayoutInflater.from(permissionsContainer.context)
        .inflate(R.layout.view_camera_permissions, permissionsContainer, true)
    view.findViewById<View>(R.id.permissions_req_btn)
        ?.setOnClickListener { requestPermissionClick() }
  }

  override fun onPause() {
    handler.removeCallbacksAndMessages(null)
    cameraView.onPause()
    super.onPause()
  }

  override fun onStop() {
    cameraView.onStop()
    super.onStop()
  }

  protected open fun requestPermissionClick() {
    requestPermissions(object : PermissionsFragment.PermissionsResultCallback {
      override fun granted() {
        cameraView.post {
          Log.d(TAG, "Permission Granted")
          cameraView.onStart()
          cameraView.onResume()
        }
      }

      override fun denied() {}
    }, *REQUIRED_PERMISSIONS)
  }

  private fun onCameraFlashClick(flashBtn: ImageView) {
    val currentFlashState = cameraView.flash
    val currentStateIndex = FLASH_STATE_CYCLE_LIST.indices.firstOrNull {
      currentFlashState == FLASH_STATE_CYCLE_LIST[it]
    } ?: 0

    val newStateIndex = (currentStateIndex + 1) % FLASH_STATE_CYCLE_LIST.size
    setFlash(flashBtn, FLASH_STATE_CYCLE_LIST[newStateIndex])
  }


  private fun onCameraFacingClick() {
    val currentFlashState = cameraView.facing
    val currentStateIndex = FACING_STATE_CYCLE_LIST.indices.firstOrNull {
      currentFlashState == FACING_STATE_CYCLE_LIST[it]
    } ?: 0

    val newStateIndex = (currentStateIndex + 1) % FACING_STATE_CYCLE_LIST.size
    setFacing(FACING_STATE_CYCLE_LIST[newStateIndex])
  }

  @UiThread
  private fun CameraKitView.onTakePhotoClick() {
    captureImage { cameraKitView, imageData ->
      Log.d(TAG, "onPictureTaken ${imageData?.size}")
      if (imageData == null) return@captureImage

      val context = cameraKitView.context ?: return@captureImage

      AsyncTask.execute {
        flexInputCoordinator?.fileManager?.newImageFile()?.also { file ->
          try {
            FileOutputStream(file).use {
              it.write(imageData)
            }

            context.addToMediaStore(file)
            cameraKitView.post {
              flexInputCoordinator?.addExternalAttachment(file.toAttachment())
            }
          } catch (e: IOException) {
            Log.w(TAG, "Cannot write to $file", e)
          }
        }
      }
    }
  }

  private fun onLaunchCameraClick() {
    val context = context ?: return

    val fileManager = flexInputCoordinator?.fileManager ?: return

    val newPhotoFile = fileManager.newImageFile()
    val photoUri = fileManager.toFileProviderUri(context, newPhotoFile)
    val takePictureIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        .putExtra(MediaStore.EXTRA_OUTPUT, photoUri)
        .addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION)

    if (takePictureIntent.resolveActivity(context.packageManager) != null) {
      this.photoFile = newPhotoFile
      context.grantWriteAccessToURI(takePictureIntent, photoUri)
      startActivityForResult(takePictureIntent, REQUEST_IMAGE_CAPTURE)
    }
  }

  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
    super.onActivityResult(requestCode, resultCode, data)
    if (REQUEST_IMAGE_CAPTURE != requestCode) {
      return
    }

    photoFile?.also {
      when (resultCode) {
        Activity.RESULT_CANCELED -> { /* Do nothing*/
        }
        Activity.RESULT_OK -> {
          context?.addToMediaStore(it)
          flexInputCoordinator?.addExternalAttachment(it.toAttachment())
        }
        else -> {
          Toast.makeText(context, R.string.camera_intent_result_error, Toast.LENGTH_SHORT).show()
          it.delete()  // cleanup
        }
      }
    }
  }

  private val cameraErrorListener =
      CameraKitView.ErrorListener { _, cameraException ->
        Log.d(TAG, "onCameraError: $cameraException")
      }

  private val cameraListener = object : CameraKitView.CameraListener {
    override fun onOpened() {
      Log.d(TAG, "onCameraOpened")
      cameraView.apply { onFacingChanged(facing) }
    }

    override fun onClosed() {
      Log.d(TAG, "onCameraClosed")
    }
  }

  private fun setFacing(newFacingState: Int) {
    cameraView.apply {
      if (facing != newFacingState) {
        facing = newFacingState
        Toast.makeText(context, R.string.camera_switched, Toast.LENGTH_SHORT).show()
      }
    }
    onFacingChanged(newFacingState)
  }

  private fun onFacingChanged(newFacingState: Int) {
    @DrawableRes val facingImg: Int =
        when (newFacingState) {
          CameraKit.FACING_FRONT -> R.drawable.ic_camera_rear_white_24dp
//          CameraView.FACING_BACK,
          else -> R.drawable.ic_camera_front_white_24dp
        }
    cameraFacingBtn.setImageResource(facingImg)
  }

  private fun setFlash(btn: ImageView, newFlashState: Int, notify: Boolean = true) {
    if (cameraView.flash == newFlashState) {
      return
    }
    flashState = newFlashState

    try {
      cameraView.flash = newFlashState
      btn.setImageResource(when (cameraView.flash) {
        CameraKit.FLASH_ON -> R.drawable.ic_flash_on_24dp
        CameraKit.FLASH_OFF -> R.drawable.ic_flash_off_24dp
        else -> R.drawable.ic_flash_auto_24dp
      })

      if (notify) onFlashChanged(btn, cameraView.flash)

    } catch (e: Exception) {
      onCameraError(e, "Camera error on set flash")
      Toast.makeText(context, R.string.camera_unknown_error, Toast.LENGTH_SHORT)
          .show()
    }
  }

  protected open fun onCameraError(e: Exception, message: String) {
    Log.e(TAG, message, e)
  }

  companion object {
    private val REQUIRED_PERMISSIONS = arrayOf(
        Manifest.permission.WRITE_EXTERNAL_STORAGE,
        Manifest.permission.READ_EXTERNAL_STORAGE,
        Manifest.permission.CAMERA)

    private val FLASH_STATE_CYCLE_LIST = intArrayOf(
        CameraKit.FLASH_AUTO,
        CameraKit.FLASH_ON,
        CameraKit.FLASH_OFF)

    private val FACING_STATE_CYCLE_LIST = intArrayOf(
        CameraKit.FACING_BACK,
        CameraKit.FACING_FRONT)

    private val TAG = CameraFragment::class.java.canonicalName

    private const val PREF_FLASH_STATE = "FLASH_STATE"

    const val REQUEST_IMAGE_CAPTURE = 4567

    /**
     * This is a hack to allow the file provider API to still
     * work on older API versions.
     *
     * http://bit.ly/2iC4bUJ
     */
    private fun Context.grantWriteAccessToURI(intent: Intent, uri: Uri) {
      val resInfoList =
          packageManager.queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY)

      for (resolveInfo in resInfoList) {
        val packageName = resolveInfo.activityInfo.packageName
        val mode = Intent.FLAG_GRANT_WRITE_URI_PERMISSION or Intent.FLAG_GRANT_READ_URI_PERMISSION

        grantUriPermission(packageName, uri, mode)
      }
    }


    private fun Context.addToMediaStore(photo: File) {
      val mediaScanIntent = Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE, Uri.fromFile(photo))
      sendBroadcast(mediaScanIntent)
      Log.d(TAG, "Photo added to MediaStore: ${photo.name}")
    }
  }
}
